codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/nameresolver/Dafny.java

```diff
@@ -92,6 +97,74 @@ public static String datatypeConstructorCreate(String name, boolean isRecordType
         return "create_" + DafnyNameResolverHelpers.dafnyCompilesExtra_(name);
     }
 
+    private static final DafnyVersion NEEDS_TYPE_DESCRIPTORS_WHEN_CONSTRUCTING_DATATYPES = new DafnyVersion(4, 2, 0);
+
+    public static boolean datatypeConstructorsNeedTypeDescriptors(DafnyVersion dafnyVersion) {
+        return dafnyVersion.compareTo(NEEDS_TYPE_DESCRIPTORS_WHEN_CONSTRUCTING_DATATYPES) >= 0;
+    }
+
+    private boolean datatypeConstructorsNeedTypeDescriptors() {
+        return datatypeConstructorsNeedTypeDescriptors(dafnyVersion);
+    }
+
+    public CodeBlock createNone(CodeBlock typeDescriptor) {
+        if (datatypeConstructorsNeedTypeDescriptors()) {
+            return CodeBlock.of(
+                    "$T.create_None($L)",
+                    ClassName.get("Wrappers_Compile", "Option"),
+                    typeDescriptor);
+        } else {
+            return CodeBlock.of(
+                    "$T.create_None()",
+                    ClassName.get("Wrappers_Compile", "Option"));
+        }
+    }
+
+    public CodeBlock createSome(CodeBlock typeDescriptor, CodeBlock value) {
+        if (datatypeConstructorsNeedTypeDescriptors()) {
+            return CodeBlock.of(
+                    "$T.create_Some($L, $L)",
+                    Constants.DAFNY_OPTION_CLASS_NAME,
+                    typeDescriptor,
+                    value);
+        } else {
+            return CodeBlock.of(
+                    "$T.create_Some($L)",
+                    Constants.DAFNY_OPTION_CLASS_NAME,
+                    value);
+        }
+    }
+
+    public CodeBlock createSuccess(CodeBlock valueTypeDescriptor, CodeBlock value) {
+        if (datatypeConstructorsNeedTypeDescriptors()) {
+            return CodeBlock.of(
+                    "$T.create_Success($L, Error._typeDescriptor(), $L)",
+                    Constants.DAFNY_RESULT_CLASS_NAME,
+                    valueTypeDescriptor,
+                    value);
+        } else {
+            return CodeBlock.of(
+                    "$T.create_Success($L)",
+                    Constants.DAFNY_RESULT_CLASS_NAME,
+                    value);
+        }
+    }
+
+    public CodeBlock createFailure(CodeBlock typeDescriptor, CodeBlock error) {
+        if (datatypeConstructorsNeedTypeDescriptors()) {
+            return CodeBlock.of(
+                    "$T.create_Failure($L, Error._typeDescriptor(), $L)",
+                    Constants.DAFNY_RESULT_CLASS_NAME,
+                    typeDescriptor,
+                    error);
+        } else {
+            return CodeBlock.of(
+                    "$T.create_Failure($L)",
+                    Constants.DAFNY_RESULT_CLASS_NAME,
+                    error);
+        }
+    }
+
```

Comment:
Blocking: Add Java docs.
Smithy-Dafny is a wildly confusing project with many hands contributing functionality.
We MUST describe what our methods do.

I checked, its true, all of these method names are unique.
But they are barely unique.

We need to java docs so future us know when to invoke this these methods.